example/reduce/src/reduce.cpp

```diff
@@ -0,0 +1,187 @@
+            devAcc, static_cast<Extent>(blockCount));
+
+    // copy the data to the GPU
+    alpaka::mem::view::copy(queue, sourceDeviceMemory, hostMemory, n);
+
+    // create kernels with their workdivs
+    ReduceKernel<blockSize, T, TFunc> kernel1, kernel2;
+    WorkDiv workDiv1{ static_cast<Extent>(blockCount),
+                      static_cast<Extent>(blockSize),
+                      static_cast<Extent>(1) };
+    WorkDiv workDiv2{ static_cast<Extent>(1),
+                      static_cast<Extent>(blockSize),
+                      static_cast<Extent>(1) };
+
+    // execute first kernel
+    auto const exec1(alpaka::kernel::createTaskExec<Acc>(
+        workDiv1,
+        kernel1,
+        alpaka::mem::view::getPtrNative(sourceDeviceMemory),
+        alpaka::mem::view::getPtrNative(destinationDeviceMemory),
+        n,
+        func));
+
+    // reduce the last block
+    auto const exec2(alpaka::kernel::createTaskExec<Acc>(
+        workDiv2,
+        kernel2,
+        alpaka::mem::view::getPtrNative(destinationDeviceMemory),
+        alpaka::mem::view::getPtrNative(destinationDeviceMemory),
+        blockCount,
+        func));
+
+    // enqueue both kernels
+    alpaka::queue::enqueue(queue, exec1);
+    alpaka::queue::enqueue(queue, exec2);
+
+    //  download result from GPU
+    T resultGpuHost;
+    auto resultGpuDevice =
+        alpaka::mem::view::ViewPlainPtr<DevHost, T, Dim, Idx>(
+            &resultGpuHost, devHost, static_cast<Extent>(blockSize));
+
+    alpaka::mem::view::copy(queue, resultGpuDevice, destinationDeviceMemory, 1);
+
+    return resultGpuHost;
+}
+
+int main()
+{
+    // select device and problem size
+    const int dev = 0;
+    uint64_t n = 1 << 28;
+
+    using T = uint32_t;
+    static constexpr uint64_t blockSize = getMaxBlockSize<Accelerator, 256>();
+    
+    DevAcc devAcc(alpaka::pltf::getDevByIdx<PltfAcc>(dev));
+    DevHost devHost(alpaka::pltf::getDevByIdx<PltfHost>(0u));
+    QueueAcc queue(devAcc);
+
+    // calculate optimal block size (8 times the MP count proved to be
+    // relatively near to peak performance in benchmarks)
+    uint32_t blockCount = static_cast<uint32_t>(
+        alpaka::acc::getAccDevProps<Acc, DevAcc>(devAcc).m_multiProcessorCount *
+        8);
+    uint32_t maxBlockCount = static_cast<uint32_t>(
+        (((n + 1) / 2) - 1) / blockSize + 1); // ceil(ceil(n/2.0)/blockSize)
+
+    if (blockCount > maxBlockCount)
+        blockCount = maxBlockCount;
+
+    // allocate memory
+    auto hostMemory = alpaka::mem::buf::alloc<T, Idx>(devHost, n);
+
+    T *nativeHostMemory = alpaka::mem::view::getPtrNative(hostMemory);
+
+    // fill array with data
+    for (uint64_t i = 0; i < n; i++)
+        nativeHostMemory[i] = static_cast<T>(i + 1);
+
+    // define the reduction function
+    auto addFn = [] ALPAKA_FN_ACC(T a, T b) -> T { return a + b; };
+
+    // reduce
+    T result = reduce<T, decltype(addFn)>(devHost, devAcc, queue, n, hostMemory, addFn);
```

Comment:
Do you really need to write the template parameters? You are providing all the parameters and the templates should be deducible.

Reply:
The ```decltype(addFn)``` seems to be deducible but it wont compile if the ```T``` in line 162 is left out es well.